### PR TITLE
Afform Entity Reference Submit

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
Improve Afform Submit handling to handle for entity reference fields that reference other form elements

Before
----------------------------------------
Cannot use a value of Individual1 for a field such as source_contact_id for an Activity 

After
----------------------------------------
Can use a value of Individual1 for field source_contact_id when creating an activity using Afform

Technical Details
----------------------------------------
This PR changes the Afform Submit Event dispatching so that it dispatches once per record / entity rather than once per afform submit

ping @colemanw @totten

### CODE REVIEW

File: ext/afform/core/CRM/Afform/Utils.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+class CRM_Afform_Utils {
```
Comment (colemanw) [inline ext/afform/core/CRM/Afform/Utils.php:18]: I suggest putting this class in the `Civi\Afform` namespace.

**Comment by colemanw** [comment]
This is shaping up nicely. Aside from a few nitpicks from the stylechecker, I just have a couple suggestions:
1. Replace entity references before firing the submit event, so the listeners don't have to do all that work. Make a function for it and that function should check getfields to validate that the field is indeed an FK before attempting any replacements.
2. Add an Afform.submit api4 test to show that a ref like `activity_target = 'Contact1` works but that `'first_name' = "Contact2"` doesn't get accidentally replaced just because it happens to be the name of an entity.
3. Out of curiosity, is there any check being done to prevent self-references? E.g. a field in `Contact1` references `Contact1.id`?

**Comment by seamuslee001** [comment]
@colemanw I have done 1. but haven't looked into 2 & 3 as yet but doing 1 prompted some fixes so might be good to review it

**Comment by seamuslee001** [comment]
@colemanw I believe Item 2 is done but not sure if item 3 is needed at this point

**Comment by colemanw** [comment]
Thanks @seamuslee001 this looks good. I did some cleanup on related areas of code and realized that there's still more as I started to uncover some shortcomings in the preprocess/submit apis but if tests are passing now then this is in a good place to be merged.

**Comment by colemanw** [comment]
@seamuslee001 here's a funny story: I added a new unit test with a more complex scenario for `Civi\Afform\Utils::getEntityWeights()`, and was struggling to get it to pass. I had a hard time grokking that function to debug it and started to wonder if there was some code snippet on StackExchange I could steal. A quick web search landed me on https://github.com/marcj/topsort.php where I thought, "this looks nice but I don't want to introduce yet another dependency." Just when I was about to look elsewhere I did a double-take that one of the contributors was @totten! And lo and behold, it's already a core dependency (he used it to sort Extensions by order of dependencies).
5 minutes later I had it dropped into the `getEntityWeights` function and the test passed :)

**Pull request merged**